### PR TITLE
Support for 'rank' question type #4594

### LINKED ISSUE #4594: Support for 'rank' question type
as per title.

Need support for allowing duplicate ranks

In future:
* Support for options with same rank (in future)
    * js prompt for user to decide behaviour when selecting the same rank for 2 options
        * to keep ranking consistent (.e.g when 2 options have rank 1, the rank-2 option should either become rank-3 or unranked)
        * 
* Display a visual of the ranked options (in future)
    * beside the options, to show the user the order of the ranked options

### PR DESCRIPTION
**Pull request by kanghj**
Fixes #4594

Added rank question type for ranking options, and ranking recipients. 2 classes: FeedbackRankOptionsQuestionDetails and FeedbackRankRecipientsQuestionDetails.

For both types, it is possible for the instructor to specify if duplicate ranks to multiple options are allowed. Currently, it is still possible for the response giver to not rank every option/recipient given, to support the main use case requested for this question type. The option to allow the instructor to specify that can be added as a separate issue after this is done. Enhancements to usability and responsiveness, such as improving the response submission form by introducing a drag-and-drop library or displaying a visualisation of the ranking, can also be done separately in other issues.

### CODE REVIEW

**Comment by damithc** [comment]
I want to use this PR as an example of how to add a new question type. As such, try to extract out any changes not strictly related to adding the new question type into separate PRs.

File: src/main/java/teammates/common/datatransfer/FeedbackQuestionType.java [hunk 1]
```diff
@@ -96,6 +104,12 @@ public FeedbackResponseDetails getFeedbackResponseDetailsInstance(FeedbackQuesti
             case RUBRIC:
                 feedbackResponseDetails = new FeedbackRubricResponseDetails();
                 break;
+            case RANK_OPTIONS:
+                feedbackResponseDetails = new FeedbackRankOptionsResponseDetails();
+                break;
+            case RANK_RECIPIENTS:
+                feedbackResponseDetails = new FeedbackRankRecipientsResponseDetails();
+                break;
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackQuestionType.java]: Any chance we can use inheritance here instead?

File: src/main/resources/feedbackQuestionRankRecipientsEditFormTemplate.html [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+        
+    </div>
+    <div class="col-sm-6">
+        <input type="checkbox" name="${Const.ParamsNames.FEEDBACK_QUESTION_RANKISDUPLICATESALLOWED}" 
+               id="${Const.ParamsNames.FEEDBACK_QUESTION_RANKISDUPLICATESALLOWED}-${questionNumber}" 
+               ${areDuplicatesAllowedChecked}>
+        <span style="margin-left: 5px; font-weight: bold;">Allow response giver to give the same rank to multiple options</span>
+    </div>
+    
+
```
Comment (damithc) [inline src/main/resources/feedbackQuestionRankRecipientsEditFormTemplate.html]: These empty lines are intentional?

**Comment by LowWeiLin** [comment]
image
When the response is incomplete or complete, there is no blue/red text, but there are 2 gray lines.
Can remove the lines when there is no text? And keep the blue text if not all ranks are assigned yet.

**Comment by LowWeiLin** [comment]
If the same rank is allowed for multiple options, what are the valid ranks?
e.g. (1, 2, 2) and (1, 3, 3) are both valid?
And how would the stats for this be calculated?

**Comment by LowWeiLin** [comment]
Also, remember to add the new question types to the "All question types" feedback session that is created for new instructors.

**Comment by damithc** [comment]
I guess user guides need to be updated too?

**Comment by kanghj** [comment]
Ok, will update them. I think both (1,2,2) and (1,3,3) should be allowed, as they are both valid ways of representing ties, but I will normalise them during the computation of statistics.

**Comment by kanghj** [comment]
Changes made

**Comment by kanghj** [comment]
@LowWeiLin, some changes made. Review this again?

**Pull request merged**